File: remote/src/main/java/com/kelsos/mbrc/presenters/AlbumTracksPresenterImpl.java
package com.kelsos.mbrc.presenters;

import com.google.inject.Inject;
import com.kelsos.mbrc.annotations.MetaDataType;
import com.kelsos.mbrc.annotations.Queue;
import com.kelsos.mbrc.domain.Track;
import com.kelsos.mbrc.interactors.AlbumTrackInteractor;
import com.kelsos.mbrc.interactors.QueueInteractor;
import com.kelsos.mbrc.ui.views.AlbumTrackView;
import roboguice.util.Ln;
import rx.android.schedulers.AndroidSchedulers;
import timber.log.Timber;

public class AlbumTracksPresenterImpl implements AlbumTracksPresenter {

  private AlbumTrackView view;
  @Inject private AlbumTrackInteractor albumTrackInteractor;
  @Inject private QueueInteractor interactor;

  @Override public void bind(AlbumTrackView view) {
    this.view = view;
  }

  @Override public void load(long albumId) {
    albumTrackInteractor.execute(albumId).observeOn(AndroidSchedulers.mainThread()).subscribe(albumTrackModel -> {
      view.updateTracks(albumTrackModel.getTracks());
      view.updateAlbum(albumTrackModel.getAlbum());
    }, Ln::v);
  }

  @Override public void play(long albumId) {
    interactor.execute(MetaDataType.ALBUM, Queue.NOW, (int) albumId).subscribe(aBoolean -> {
      if (aBoolean) {
        view.showPlaySuccess();
      } else {
        view.showPlayFailed();
      }
    }, t -> {
      view.showPlayFailed();
      Timber.e(t, "Failed to play the album %d", albumId);
    });
  }

  @Override public void queue(Track entry, @Queue.Action String action) {
    final long id = entry.getId();
    interactor.execute(MetaDataType.TRACK, action, (int) id).subscribe(aBoolean -> {
      if (aBoolean) {
        view.showTrackSuccess();
      } else {
        view.showTrackFailed();
      }
    }, t -> {
      view.showTrackFailed();
      Timber.e(t, "Failed to play the track %d", id);
    });
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/presenters/AlbumTracksPresenterImpl.kt
package com.kelsos.mbrc.presenters

import com.kelsos.mbrc.annotations.MetaDataType
import com.kelsos.mbrc.annotations.Queue
import com.kelsos.mbrc.domain.Track
import com.kelsos.mbrc.interactors.AlbumTrackInteractor
import com.kelsos.mbrc.interactors.QueueInteractor
import com.kelsos.mbrc.ui.views.AlbumTrackView
import rx.android.schedulers.AndroidSchedulers
import timber.log.Timber
import javax.inject.Inject

class AlbumTracksPresenterImpl
@Inject constructor(private val albumTrackInteractor: AlbumTrackInteractor,
                    private val interactor: QueueInteractor): AlbumTracksPresenter {

  private var view: AlbumTrackView? = null

  override fun bind(view: AlbumTrackView) {
    this.view = view
  }

  override fun load(albumId: Long) {
    albumTrackInteractor.execute(albumId).observeOn(AndroidSchedulers.mainThread()).subscribe({
      view?.updateTracks(it.tracks)
      view?.updateAlbum(it.album)
    }, { Timber.v(it) })
  }

  override fun play(albumId: Long) {
    interactor.execute(MetaDataType.ALBUM, Queue.NOW, albumId).subscribe({ aBoolean ->
      if (aBoolean!!) {
        view?.showPlaySuccess()
      } else {
        view?.showPlayFailed()
      }
    }) {
      view?.showPlayFailed()
      Timber.e(it, "Failed to play the album %d", albumId)
    }
  }

  override fun queue(entry: Track, @Queue.Action action: String) {
    val id = entry.id
    interactor.execute(MetaDataType.TRACK, action, id).subscribe({
      if (it!!) {
        view?.showTrackSuccess()
      } else {
        view?.showTrackFailed()
      }
    }) { t ->
      view?.showTrackFailed()
      Timber.e(t, "Failed to play the track %d", id)
    }
  }
}
